NIFI-9304 - Adding PublishGCPubSubLite and ConsumeGCPubSubLite processors

pvillard31 opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Adding PublishGCPubSubLite and ConsumeGCPubSubLite processors

This pull request is adding two processors to interact with GCP Pub/Sub Lite. I initially thought about improving the existing processors but since it uses quite different APIs, I decided to go with new processors.

I did some extensive testing in a GCP environment and it seems to be working as expected including under high load and with concurrent tasks.

Please note that I had to use the compute engine credentials or the application credentials. If providing a service account JSON payload/file in the controller service, it seems that some gRPC calls are ignoring this authentication method. It seems to be specific to Pub/Sub Lite as I was able to use the GCS, PubSub and BigQuery processors using the service account JSON file.

Some screenshots of my tests:

<img width="1417" alt="Screenshot 2021-10-16 at 14 14 57" src="https://user-images.githubusercontent.com/11541012/137587930-f45c8304-6aea-444f-8a13-b000169f0fb4.png">

<img width="1782" alt="Screenshot 2021-10-16 at 14 16 52" src="https://user-images.githubusercontent.com/11541012/137587934-a5bf6886-d0b8-4f75-8bc1-68aa2cddd490.png">

<img width="1785" alt="Screenshot 2021-10-16 at 14 17 06" src="https://user-images.githubusercontent.com/11541012/137587939-be564e74-61ff-477e-88c2-1eee2bfb9546.png">

<img width="756" alt="Screenshot 2021-10-16 at 14 18 03" src="https://user-images.githubusercontent.com/11541012/137587947-f7c63425-fe65-496f-adc4-ddd4cebdc533.png">

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/ConsumeGCPubSubLite.java:
```diff
@@ -0,0 +1,258 @@
+
+        final Map<String, String> attributes = new HashMap<>();
+        attributes.put(MESSAGE_ID_ATTRIBUTE, message.getMessage().getMessageId());
+        attributes.put(ORDERING_KEY_ATTRIBUTE, message.getMessage().getOrderingKey());
```
Just to confirm, this would only be populated if someone set the ordering key by publishing external to NiFi?  I couldn't find a way to set the ordering key on the `PublishGCPubSubLite` processor, but did see it being set when publishing using `gcloud`.  Perhaps a feature for another day.

gresockj reviewed (commented):
This works well, and drastically outperformed the standard `PublishGCPubSub`/`ConsumeGCPubSub` when I tried it out.

One additional note -- it appears that the gRCP API is not using the credentials provided by the controller service at all if you are not running within the Compute Engine.  If you are, either `Use Application Default Credentials` or `Use Compute Engine Credentials` will work.  I don't think this should block the PR, but it should be noted as a current limitation in the processor documentation.

pvillard31 commented:
Thanks for the review @gresockj ! I think I addressed all of your comments!

Note that I had some weird return line characters and used dos2unix (I recently had to setup again all of my dev env...) and it was showing all lines as modified so I rebased, squash and forced push. Sorry about that...

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/ConsumeGCPubSubLite.java:
```diff
@@ -0,0 +1,284 @@
+        final List<ConfigVerificationResult> verificationResults = new ArrayList<>();
+        try {
+            if (subscriber == null) {
+                subscriber = getSubscriber(context);
```
Rather than setting the actual instance variable `subscriber`, we should simply call the `getSubscriber` method and catch any exceptions.  We don't want the verification process to actually modify the class.

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/ConsumeGCPubSubLite.java:
```diff
@@ -0,0 +1,284 @@
+    }
+
+    @Override
+    public List<ConfigVerificationResult> verify(ProcessContext context, ComponentLog verificationLogger, Map<String, String> attributes) {
```
`final` arguments

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/ConsumeGCPubSubLite.java:
```diff
@@ -0,0 +1,284 @@
+            verificationResults.add(new ConfigVerificationResult.Builder()
+                    .verificationStepName("Creation of the Subscriber")
+                    .outcome(Outcome.FAILED)
+                    .explanation("Failed to create Google Cloud PubSub Lite Subscriber: " + e)
```
I think this will need to be `e.getMessage()`

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/ConsumeGCPubSubLite.java:
```diff
@@ -0,0 +1,284 @@
+            getLogger().error("Failed to create Google Cloud PubSub Lite Subscriber", e);
+
+            verificationResults.add(new ConfigVerificationResult.Builder()
+                    .verificationStepName("Creation of the Subscriber")
```
The grammatical style we've been using elsewhere is: "Create the Subscriber"

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/ConsumeGCPubSubLite.java:
```diff
@@ -0,0 +1,284 @@
+        try {
+            if (subscriber == null) {
+                subscriber = getSubscriber(context);
+            }
```
Here we should add a verification result with a successful outcome to show that the subscriber was able to be created.

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/ConsumeGCPubSubLite.java:
```diff
@@ -0,0 +1,284 @@
+@InputRequirement(InputRequirement.Requirement.INPUT_FORBIDDEN)
+@Tags({"google", "google-cloud", "gcp", "message", "pubsub", "consume", "lite"})
+@CapabilityDescription("Consumes message from the configured Google Cloud PubSub Lite subscription. In its current state, this processor "
+        + "will work only if using the GCP Credentials Controller Service with 'Use Application Default Credentials' or 'Use Compute Engine Credentials'.")
```
I think the important part to add is that it only works on a Compute Engine instance, because if you try "Use Application Default Credentials" without running on Compute, it still fails.

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/PublishGCPubSubLite.java:
```diff
@@ -0,0 +1,326 @@
+
+    @Override
+    public List<ConfigVerificationResult> verify(ProcessContext context, ComponentLog verificationLogger, Map<String, String> attributes) {
+        final List<ConfigVerificationResult> verificationResults = new ArrayList<>();
```
See same notes as in `ConsumeGCPubSubLite`

gresockj reviewed (changes requested):
Thanks for the updates, @pvillard31!  Just a few more items left.

gresockj reviewed (commented):
Great, that addresses all my comments, @pvillard31.  I plan to run this with the config verification updates, and if all looks good, I'll be a +1.

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/PublishGCPubSubLite.java:
```diff
@@ -0,0 +1,330 @@
+                final String orderingKey = context.getProperty(ORDERING_KEY).evaluateAttributeExpressions(flowFile).getValue();
+
+                PubsubMessage message= null;
+                if (orderingKey != null) {
+                    message = PubsubMessage.newBuilder().setData(flowFileContent)
+                            .setOrderingKey(orderingKey)
+                            .setPublishTime(Timestamp.newBuilder().build())
+                            .putAllAttributes(getDynamicAttributesMap(context, flowFile))
+                            .build();
+                } else {
+                    message = PubsubMessage.newBuilder().setData(flowFileContent)
+                            .setPublishTime(Timestamp.newBuilder().build())
+                            .putAllAttributes(getDynamicAttributesMap(context, flowFile))
+                            .build();
+                }
```
This can be streamlined a little by declaring the builder as a final local variable, and then adding the orderingKey if it's non-null.

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/PublishGCPubSubLite.java:
```diff
@@ -0,0 +1,330 @@
+                final ByteString flowFileContent = ByteString.copyFrom(baos.toByteArray());
+                final String orderingKey = context.getProperty(ORDERING_KEY).evaluateAttributeExpressions(flowFile).getValue();
+
+                PubsubMessage message= null;
```
Can be final, also missing a space

gresockj commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/lite/PublishGCPubSubLite.java:
```diff
@@ -0,0 +1,330 @@
+            .build();
+
+    public static final PropertyDescriptor ORDERING_KEY = new PropertyDescriptor
+            .Builder().name("gcp-ordering_key")
```
Perhaps `gcp-ordering-key` instead of `gcp-ordering_key`

gresockj reviewed (changes requested):
Ok, config verification looks good, but I detected a couple minor changes needed for the new ORDERING_KEY support on the publisher.

pvillard31 commented:
Thanks @gresockj , good points, I think this should be OK now.

gresockj reviewed (approved):
Thanks for these last updates, @pvillard31!  I've run this again on Compute Engine, and it all works as expected now.  +1 from me, though I am going to wait before merging until after the 1.15.0 release.
